File: app/src/main/java/seigneur/gauvain/mycourt/ui/pin/PinViewModel.java
package seigneur.gauvain.mycourt.ui.pin;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import seigneur.gauvain.mycourt.data.model.Pin;
import seigneur.gauvain.mycourt.data.repository.UserRepository;
import seigneur.gauvain.mycourt.ui.pin.tasks.CheckCurrentPinTask;
import seigneur.gauvain.mycourt.ui.pin.tasks.StorePinTask;
import seigneur.gauvain.mycourt.utils.Constants;
import seigneur.gauvain.mycourt.utils.SingleLiveEvent;
import seigneur.gauvain.mycourt.utils.crypto.DeCryptor;
import seigneur.gauvain.mycourt.utils.crypto.EnCryptor;
import timber.log.Timber;

public class PinViewModel extends ViewModel implements
        CheckCurrentPinTask.CheckCurrentPinCallBack, StorePinTask.StorePinCallback {

    @Inject
    public UserRepository mUserRepository;

    @Inject
    EnCryptor mEnCryptor;

    @Inject
    DeCryptor mDeCryptor;

    private CompositeDisposable mCompositeDisposable = new CompositeDisposable();

    private Pin mStoredPin = null; //not related to UI
    private MutableLiveData<Integer> mPinstep = new MutableLiveData<>(); //related to UI
    private SingleLiveEvent<String> mConfirmCurrentPin = new SingleLiveEvent<>(); //related to UI

    private byte[] iV; //do not set it as LiveData, we don't want to share it in the UI
    private boolean isDeCryptorEnabled=false; //do not set it as LiveData, we don't want to share it in the UI

    //TODO - must compare string between step 0 an step 1 before register pin

    private CheckCurrentPinTask mCheckCurrentPinTask;
    private StorePinTask mStorePinTask;

    @Inject
    public PinViewModel() {}

    public void init() {
        initTasks();
        //Do not perform checking request if stored pin is already defined
        if (mStoredPin!=null)
            return;
        mCheckCurrentPinTask.checkIfPinAlreadyExists();
    }

    private void initTasks() {
        if (mCheckCurrentPinTask==null)
            mCheckCurrentPinTask = new CheckCurrentPinTask(mCompositeDisposable,
                    mUserRepository, mDeCryptor, this);

        if (mStorePinTask==null)
            mStorePinTask = new StorePinTask(mCompositeDisposable,
                    mUserRepository, mEnCryptor, this);
    }

    public LiveData<Integer> getStep() {
        return mPinstep;
    }

    @Override
    public void onCleared(){
        super.onCleared();
        Timber.d("viewmodel cleared");
        mCompositeDisposable.clear();
    }

    /**
     * If a stored pin exists, ask user to confirm it in order
     * to allow him to redefine it
     */
    public void onCurrentPinConfirmed(String pinInput) {
        Timber.d("onCurrentPinConfirmed");
        //comparePinAndInput(pinInput);
        mCheckCurrentPinTask.comparePinAndInput(mStoredPin, pinInput);
    }

    /**
     *
     */
    public void onFirstPinDefined(String pin) {
        mPinstep.setValue(Constants.PIN_STEP_NEW_PIN_TWO); //set current step to 1
        //todo - live data single event
        /*if (mPinView!=null) {
            mPinView.showCreationPinStep(1);
        }*/
    }

    public void onNewPinConfirmed(String pin) {
        mStorePinTask.cryptAndStorePin(pin);
    }

    /*
    *******************************************************************************
    * CheckCurrentPinTask.CheckCurrentPinCallBack
    *******************************************************************************/
    @Override
    public void onPinAlreadyStored(Pin pin) {
        mStoredPin = pin;
        mPinstep.setValue(Constants.PIN_STEP_CHECK_STORED); //set current step to -1
    }

    @Override
    public void onNoPinStored() {
        mPinstep.setValue(Constants.PIN_STEP_NEW_PIN_ONE);
    }

    @Override
    public void onPinCheckSuccess() {

    }

    @Override
    public void onCheckPinFailed() {

    }

    /*
    *******************************************************************************
    * StorePinTask.StorePinCallback
    *******************************************************************************/
    @Override
    public void onPinStored() {

    }

    @Override
    public void onStorePinFailed() {

    }

}

File: app/src/main/java/seigneur/gauvain/mycourt/ui/pin/PinViewModel.kt
package seigneur.gauvain.mycourt.ui.pin

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel

import javax.inject.Inject

import io.reactivex.disposables.CompositeDisposable
import seigneur.gauvain.mycourt.data.model.Pin
import seigneur.gauvain.mycourt.data.repository.UserRepository
import seigneur.gauvain.mycourt.ui.pin.tasks.CheckCurrentPinTask
import seigneur.gauvain.mycourt.ui.pin.tasks.StorePinTask
import seigneur.gauvain.mycourt.utils.Constants
import seigneur.gauvain.mycourt.utils.SingleLiveEvent
import seigneur.gauvain.mycourt.utils.crypto.DeCryptor
import seigneur.gauvain.mycourt.utils.crypto.EnCryptor
import timber.log.Timber

class PinViewModel @Inject
constructor() : ViewModel(), CheckCurrentPinTask.CheckCurrentPinCallBack, StorePinTask.StorePinCallback {

    @Inject
    lateinit var mUserRepository: UserRepository

    @Inject
    lateinit var mEnCryptor: EnCryptor

    @Inject
    lateinit var mDeCryptor: DeCryptor

    private val mCompositeDisposable = CompositeDisposable()

    private var mStoredPin: Pin? = null //not related to UI
    private val mPinstep = MutableLiveData<Int>() //related to UI
    private val mConfirmCurrentPin = SingleLiveEvent<String>() //related to UI

    private val iV: ByteArray? = null //do not set it as LiveData, we don't want to share it in the UI
    private val isDeCryptorEnabled = false //do not set it as LiveData, we don't want to share it in the UI

    //TODO - must compare string between step 0 an step 1 before register pin

    private val mCheckCurrentPinTask: CheckCurrentPinTask by lazy {
        CheckCurrentPinTask(mCompositeDisposable,
            mUserRepository, mDeCryptor, this) }

    private val mStorePinTask: StorePinTask by lazy {
        StorePinTask(mCompositeDisposable,
                mUserRepository, mEnCryptor, this)
    }

    val step: LiveData<Int>
        get() = mPinstep

    fun init() {
        //Do not perform checking request if stored pin is already defined
        if (mStoredPin != null)
            return
        mCheckCurrentPinTask.checkIfPinAlreadyExists()
    }


    public override fun onCleared() {
        super.onCleared()
        Timber.d("viewmodel cleared")
        mCompositeDisposable.clear()
    }

    /**
     * If a stored pin exists, ask user to confirm it in order
     * to allow him to redefine it
     */
    fun onCurrentPinConfirmed(pinInput: String) {
        Timber.d("onCurrentPinConfirmed")
        //comparePinAndInput(pinInput);
        mCheckCurrentPinTask.comparePinAndInput(mStoredPin!!, pinInput)
    }

    /**
     *
     */
    fun onFirstPinDefined(pin: String) {
        mPinstep.value = Constants.PIN_STEP_NEW_PIN_TWO //set current step to 1
        //todo - live data single event
        /*if (mPinView!=null) {
            mPinView.showCreationPinStep(1);
        }*/
    }

    fun onNewPinConfirmed(pin: String) {
        mStorePinTask.cryptAndStorePin(pin)
    }

    /*
    *******************************************************************************
    * CheckCurrentPinTask.CheckCurrentPinCallBack
    *******************************************************************************/
    override fun onPinAlreadyStored(pin: Pin) {
        mStoredPin = pin
        mPinstep.value = Constants.PIN_STEP_CHECK_STORED //set current step to -1
    }

    override fun onNoPinStored() {
        mPinstep.value = Constants.PIN_STEP_NEW_PIN_ONE
    }

    override fun onPinCheckSuccess() {

    }

    override fun onCheckPinFailed() {

    }

    /*
    *******************************************************************************
    * StorePinTask.StorePinCallback
    *******************************************************************************/
    override fun onPinStored() {

    }

    override fun onStorePinFailed() {

    }

}
